Mathlib/RingTheory/Regular/RegularSequence.lean

```diff
@@ -0,0 +1,645 @@
+
+end Definitions
+
+section Congr
+
+variable [CommRing R] [CommRing S] [AddCommGroup M]
+    [AddCommGroup M₂] [Module R M] [Module R M₂] [Module S M₂]
+    {σ : R →+* S} {σ' : S →+* R} [RingHomInvPair σ σ'] [RingHomInvPair σ' σ]
+
+open DistribMulAction AddSubgroup in
+lemma _root_.AddEquiv.isWeaklyRegular_congr {e : M ≃+ M₂} {as bs}
+    (h : List.Forall₂ (fun (r : R) (s : S) => ∀ x, e (r • x) = s • e x) as bs) :
+    IsWeaklyRegular M as ↔ IsWeaklyRegular M₂ bs :=
+  have H₁ i := by
+    replace h := List.forall₂_take i h
+    generalize as.take i = as' at h ⊢; generalize bs.take i = bs' at h ⊢
+    induction h with
+    | nil =>
+      convert AddSubgroup.map_bot (e : M →+ M₂) using 1 <;>
+        rw [Ideal.ofList_nil, bot_smul, bot_toAddSubgroup]
+    | @cons r s _ _ h _ ih =>
+      conv => congr <;> rw [Ideal.ofList_cons, sup_smul, sup_toAddSubgroup,
+        ideal_span_singleton_smul, pointwise_smul_toAddSubgroup,
+        top_toAddSubgroup, pointwise_smul_def]
+      apply DFunLike.ext ((e : M →+ M₂).comp (toAddMonoidEnd R M r))
+        ((toAddMonoidEnd S M₂ s).comp e) at h
+      rw [AddSubgroup.map_sup, ih, map_map, h, ← map_map,
+        map_top_of_surjective e e.surjective]
+  have H₂ i :=
+    let e' :
+       (M ⧸ (Ideal.ofList (as.take i) • ⊤ : Submodule R M)) ≃+
+              M₂ ⧸ (Ideal.ofList (bs.take i) • ⊤ : Submodule S M₂) :=
+      QuotientAddGroup.congr _ _ e (H₁ i)
+    forall_prop_congr
+      (q' := fun h' => have := h.length_eq ▸ h'; IsSMulRegular _ bs[i])
+      (fun _ => e'.isSMulRegular_congr <| (mkQ_surjective _).forall.mpr (by
+        dsimp only [e']; refine congrArg (mkQ _) ?_; exact h.get _ _ ·))
+      (iff_of_eq (congrArg _ h.length_eq))
+  (isWeaklyRegular_iff _ _).trans <|
+    (forall_congr' H₂).trans (isWeaklyRegular_iff _ _).symm
+
+lemma _root_.LinearEquiv.isWeaklyRegular_congr' (e : M ≃ₛₗ[σ] M₂) (rs : List R) :
+    IsWeaklyRegular M rs ↔ IsWeaklyRegular M₂ (rs.map σ) :=
+  e.toAddEquiv.isWeaklyRegular_congr <| List.forall₂_map_right_iff.mpr <|
+    List.forall₂_same.mpr fun r _ x => e.map_smul' r x
+
+lemma _root_.LinearEquiv.isWeaklyRegular_congr (e : M ≃ₗ[R] M₂) (rs : List R) :
+    IsWeaklyRegular M rs ↔ IsWeaklyRegular M₂ rs :=
+  Iff.trans (e.isWeaklyRegular_congr' rs) <| iff_of_eq <| congrArg _ rs.map_id
+
+end Congr
+
+lemma isWeaklyRegular_map_algebraMap_iff {R : Type*} (S M : Type*) [CommRing R]
+    [CommRing S] [Algebra R S] [AddCommGroup M] [Module R M] [Module S M]
+    [IsScalarTower R S M] (rs : List R) :
+    IsWeaklyRegular M (rs.map (algebraMap R S)) ↔ IsWeaklyRegular M rs :=
+  (AddEquiv.refl M).isWeaklyRegular_congr <| List.forall₂_map_left_iff.mpr <|
+    List.forall₂_same.mpr fun r _ => algebraMap_smul S r
+
+namespace IsWeaklyRegular
+
+variable (M) [CommRing R] [AddCommGroup M] [Module R M]
+
+variable (R) in
+lemma nil : IsWeaklyRegular M ([] : List R) :=
+  .mk (False.elim <| Nat.not_lt_zero · ·)
+
+lemma isWeaklyRegular_cons_iff (r : R) (rs : List R) :
```

Comment:
I think that all these names that start with `isWeaklyRegular` can just be in the parent namespace. Otherwise the fully-qualified names become really long, and have redundancy.

Reply:
Fwiw the fully qualified names are already super long, since they live in the namespace `RingTheory.Sequence` (I chose this namespace because `Sequence` is too ambiguous on its own). When writing this I was thinking a client could just open `IsWeaklyRegular`, but the redundancy is kinda silly. I would've defaulted to naming things like this `IsWeaklyRegular.cons_iff` but the `@[mk_iff]` naming scheme made me think this was preferred. I'll move stuff out of the `IsWeaklyRegular`/`IsRegular` namespaces as appropriate.

Also, is `RingTheory.Sequence.IsRegular` an okay name if we already have the global name `IsRegular`? I would prefer to put `Sequence` in the namespace rather than the name, but maybe it should be renamed to `IsRegularSequence` and `IsWeaklyRegular` renamed to `IsWeaklyRegularSequence` (both in the global namespace)? This seems like an annoyingly long name though

Reply:
Maybe `IsRegularSeq`?

Reply:
And it is not uncommon for mathlib to use `W` for weak/weakly (e.g. `WSeq`, `WCovBy`) so maybe `IsWRegularSeq` is also good? (less sure about this one)

Reply:
Another choice is to have `List.IsRegular` so you can write `rs.IsRegular M`.

Reply:
I initially wanted to have something like `abbrev RingTheory.Sequence R := List R` and define `Sequence.IsRegular`, but for some reason it was causing me problems. It seemed like a bad idea to introduce a superfluous abbreviaton for lists just because mathematicians call lists sequences, so I settled on putting it in the namespace

Reply:
I think also `IsRegular` is way too generic to put in the list namespace. The existing global `IsRegular` should probably be in a namespace too, imo. Why should "regular" unqualified mean this and not regularity of a sequence, of a Noetherian ring, of a noncommutative ring (vonn neumann regularity), or any of the million other predicates that mathematicians call regular?

Reply:
I am happy with the current choice of names. But I also like the `Is(W)RegularSeq` suggestion. I'll leave it up to you.

bors d+

Reply:
I think at the moment I'm going to keep it as is but will keep an eye on it going forward. It's possible that we'll quickly run into headaches where we want to talk about both kinds of `IsRegular` eg when talking about the depth of a ring